core/trino-main/src/test/java/io/trino/operator/TestDeduplicationExchangeClientBuffer.java

```diff
@@ -0,0 +1,443 @@
+                ImmutableListMultimap.of(
+                        createTaskId(0, 0),
+                        createPage("p0a0v0"),
+                        createTaskId(1, 0),
+                        createPage("p1a0v0"),
+                        createTaskId(0, 1),
+                        createPage("p0a1v0")),
+                ImmutableMap.of(
+                        createTaskId(0, 1),
+                        error),
+                ImmutableList.of("p0a1v0"))).isEqualTo(error);
+    }
+
+    private void testPollPages(Multimap<TaskId, SerializedPage> pages, Map<TaskId, RuntimeException> failures, List<String> expectedValues)
+    {
+        try (ExchangeClientBuffer buffer = new DeduplicationExchangeClientBuffer(directExecutor(), ONE_KB, RetryPolicy.QUERY)) {
+            for (TaskId taskId : Sets.union(pages.keySet(), failures.keySet())) {
+                buffer.addTask(taskId);
+            }
+            for (Map.Entry<TaskId, SerializedPage> page : pages.entries()) {
+                buffer.addPages(page.getKey(), ImmutableList.of(page.getValue()));
+            }
+            for (Map.Entry<TaskId, RuntimeException> failure : failures.entrySet()) {
+                buffer.taskFailed(failure.getKey(), failure.getValue());
+            }
+            for (TaskId taskId : Sets.difference(pages.keySet(), failures.keySet())) {
+                buffer.taskFinished(taskId);
+            }
+            buffer.noMoreTasks();
+            assertContainsPages(buffer, expectedValues);
+        }
+    }
+
+    private static void assertContainsPages(ExchangeClientBuffer buffer, List<String> values)
+    {
+        List<String> actual = new ArrayList<>();
+        while (true) {
+            SerializedPage page = buffer.pollPage();
+            if (page == null) {
+                break;
+            }
+            actual.add(page.getSlice().toStringUtf8());
+        }
+        assertTrue(buffer.isFinished());
+        assertThat(actual).containsExactlyInAnyOrderElementsOf(values);
+    }
+
+    @Test
+    public void testRemovePagesForPreviousAttempts()
+    {
+        try (ExchangeClientBuffer buffer = new DeduplicationExchangeClientBuffer(directExecutor(), ONE_KB, RetryPolicy.QUERY)) {
+            assertEquals(buffer.getRetainedSizeInBytes(), 0);
+
+            TaskId task1 = createTaskId(0, 0);
+            TaskId task2 = createTaskId(1, 0);
+            TaskId task3 = createTaskId(0, 1);
+
+            SerializedPage page1 = createPage("textofrandomlength");
+            SerializedPage page2 = createPage("textwithdifferentlength");
+            SerializedPage page3 = createPage("smalltext");
+
+            buffer.addTask(task1);
+            buffer.addPages(task1, ImmutableList.of(page1));
+            buffer.addTask(task2);
+            buffer.addPages(task2, ImmutableList.of(page2));
+
+            assertEquals(buffer.getRetainedSizeInBytes(), page1.getRetainedSizeInBytes() + page2.getRetainedSizeInBytes());
```

Comment:
can we use constants instead of `pageX.getRetaindedSizeInBytes` here? The tests assumes `pageX.getRetaindedSizeInBytes` returns actual page size. While test would still pass fine if page cleaning would not work at all, if `pageX.getRetaindedSizeInBytes` was broken and always returned `0`.

Reply:
On the other hand if constants are hardcoded it will be required to adjust the constants every time the layout of a SerializedPage changes. Also it might be more difficult to read. What do you think about adding a `assertThat(buffer.getRetainedSizeInBytes()).isGreaterThan(0);` check?

Reply:
>assertThat(buffer.getRetainedSizeInBytes()).isGreaterThan(0); check as an alternative?
that is fine.